File: backend/schemas/placement.py
# =============================================================
# schemas/placement.py — Placement Module Pydantic Schemas
# =============================================================

from __future__ import annotations
from pydantic import BaseModel, ConfigDict, field_validator
from typing import Optional, List
from datetime import datetime

from backend.models.enums import ApplicationStatus


# ---------------------------------------------------------------
# JOB POSTING SCHEMAS
# ---------------------------------------------------------------

class JobPostingCreate(BaseModel):
    company_name: str
    role_title: str
    description: Optional[str] = None
    location: Optional[str] = None
    package_lpa: Optional[float] = None
    allowed_departments: Optional[str] = None   # "cse,ece,it" or null for all
    min_cgpa: float = 0.0
    min_attendance_pct: float = 0.0
    application_deadline: Optional[datetime] = None

    @field_validator("min_cgpa")
    @classmethod
    def validate_cgpa(cls, v: float) -> float:
        if not (0.0 <= v <= 10.0):
            raise ValueError("min_cgpa must be between 0.0 and 10.0")
        return v

    @field_validator("min_attendance_pct")
    @classmethod
    def validate_attendance(cls, v: float) -> float:
        if not (0.0 <= v <= 100.0):
            raise ValueError("min_attendance_pct must be between 0 and 100")
        return v


class JobPostingUpdate(BaseModel):
    company_name: Optional[str] = None
    role_title: Optional[str] = None
    description: Optional[str] = None
    location: Optional[str] = None
    package_lpa: Optional[float] = None
    allowed_departments: Optional[str] = None
    min_cgpa: Optional[float] = None
    min_attendance_pct: Optional[float] = None
    application_deadline: Optional[datetime] = None
    is_open: Optional[bool] = None
    is_active: Optional[bool] = None


class JobPostingResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    company_name: str
    role_title: str
    description: Optional[str] = None
    location: Optional[str] = None
    package_lpa: Optional[float] = None
    allowed_departments: Optional[str] = None
    min_cgpa: float
    min_attendance_pct: float
    application_deadline: Optional[datetime] = None
    is_active: bool
    is_open: bool
    created_at: Optional[datetime] = None
    total_applications: Optional[int] = None
    is_eligible: Optional[bool] = None       # Populated when student views listing


# ---------------------------------------------------------------
# APPLICATION SCHEMAS
# ---------------------------------------------------------------

class ApplicationCreate(BaseModel):
    job_posting_id: int


class ApplicationStatusUpdate(BaseModel):
    status: ApplicationStatus
    remarks: Optional[str] = None


class ApplicationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    student_id: int
    job_posting_id: int
    status: ApplicationStatus
    remarks: Optional[str] = None
    applied_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    # Denormalised fields
    company_name: Optional[str] = None
    role_title: Optional[str] = None
    package_lpa: Optional[float] = None
    roll_number: Optional[str] = None
    student_name: Optional[str] = None


# ---------------------------------------------------------------
# ANALYTICS SCHEMAS
# ---------------------------------------------------------------

class PlacementFunnelStats(BaseModel):
    total_applied: int
    under_review: int
    shortlisted: int
    selected: int
    rejected: int
    withdrawn: int


class DepartmentPlacementStats(BaseModel):
    department: str
    total_students: int
    placed_count: int
    placement_rate: float
    avg_package_lpa: Optional[float] = None
    highest_package_lpa: Optional[float] = None


class CompanyStats(BaseModel):
    company_name: str
    total_openings: int
    total_applications: int
    students_placed: int


class PlacementAnalyticsResponse(BaseModel):
    total_job_postings: int
    active_postings: int
    total_applications: int
    total_placed_students: int
    overall_placement_rate: float
    avg_package_lpa: Optional[float] = None
    highest_package_lpa: Optional[float] = None
    funnel: PlacementFunnelStats
    by_department: List[DepartmentPlacementStats]
    top_companies: List[CompanyStats]

File: frontend/src/types/placement.ts
// ============================================================
// types/placement.ts — Placement Module Type Contracts
// ============================================================
// Every interface mirrors the corresponding backend Pydantic
// schema in backend/schemas/placement.py exactly.
// ============================================================

// ── Enums ────────────────────────────────────────────────────
export type ApplicationStatus =
  | 'applied'
  | 'under_review'
  | 'shortlisted'
  | 'selected'
  | 'rejected'
  | 'withdrawn';

// ── Job Posting ───────────────────────────────────────────────
export interface JobPosting {
  id:                   number;
  company_name:         string;
  role_title:           string;
  description:          string | null;
  location:             string | null;
  package_lpa:          number | null;
  allowed_departments:  string | null;  // "cse,ece,it"  null = all
  min_cgpa:             number;
  min_attendance_pct:   number;
  is_active:            boolean;
  is_open:              boolean;
  application_deadline: string | null;  // ISO datetime string
  created_at:           string | null;
  total_applications:   number | null;
  is_eligible:          boolean | null; // null for non-student viewers
}

export interface JobPostingCreate {
  company_name:         string;
  role_title:           string;
  description:          string;
  location:             string;
  package_lpa:          number | null;
  allowed_departments:  string | null;
  min_cgpa:             number;
  min_attendance_pct:   number;
  application_deadline: string | null;
}

// ── Application ───────────────────────────────────────────────
export interface PlacementApplication {
  id:             number;
  student_id:     number;
  job_posting_id: number;
  status:         ApplicationStatus;
  remarks:        string | null;
  applied_at:     string | null;
  updated_at:     string | null;
  // Denormalised join fields
  company_name:   string | null;
  role_title:     string | null;
  package_lpa:    number | null;
  roll_number:    string | null;
  student_name:   string | null;
}

export interface ApplicationStatusUpdate {
  status:  ApplicationStatus;
  remarks: string | null;
}

// ── Analytics ─────────────────────────────────────────────────
export interface PlacementFunnelStats {
  total_applied: number;
  under_review:  number;
  shortlisted:   number;
  selected:      number;
  rejected:      number;
  withdrawn:     number;
}

export interface DepartmentPlacementStats {
  department:          string;
  total_students:      number;
  placed_count:        number;
  placement_rate:      number;
  avg_package_lpa:     number | null;
  highest_package_lpa: number | null;
}

export interface CompanyStats {
  company_name:       string;
  total_openings:     number;
  total_applications: number;
  students_placed:    number;
}

export interface PlacementAnalytics {
  total_job_postings:      number;
  active_postings:         number;
  total_applications:      number;
  total_placed_students:   number;
  overall_placement_rate:  number;
  avg_package_lpa:         number | null;
  highest_package_lpa:     number | null;
  funnel:                  PlacementFunnelStats;
  by_department:           DepartmentPlacementStats[];
  top_companies:           CompanyStats[];
}

// ── Frontend filter + derived types ───────────────────────────
export interface PlacementFilters {
  search:           string;
  eligibleOnly:     boolean;
  minPackage:       number | '';
  maxPackage:       number | '';
  department:       string;   // "" = all
}

export const EMPTY_PLACEMENT_FILTERS: PlacementFilters = {
  search: '', eligibleOnly: false, minPackage: '', maxPackage: '', department: '',
};

// Student summary stats derived from applications
export interface StudentPlacementStats {
  total:       number;
  applied:     number;
  shortlisted: number;
  selected:    number;
  rejected:    number;
  withdrawn:   number;
}

// Departments enum values (must match backend enums.py)
export const DEPARTMENTS = [
  { value: 'cse',   label: 'CSE'   },
  { value: 'ece',   label: 'ECE'   },
  { value: 'mech',  label: 'Mech'  },
  { value: 'civil', label: 'Civil' },
  { value: 'eee',   label: 'EEE'   },
  { value: 'it',    label: 'IT'    },
  { value: 'aids',  label: 'AI&DS' },
] as const;
